### DISCUSSION TITLE
Resource account issues (Move abort in 0x1::resource_account: 0x60001)

### DISCUSSION THREAD

**Discussion by davd-ops** [aptos-labs/aptos-developer-discussions#425] [Questions]
### Discord user ID

davidbailey.eth

### Describe your question in detail.

Hello, I have [this code](https://github.com/davd-ops/aptos.institute/blob/main/sources/AptosInstitute.move). Im having problems with publishing it on testnet.

When I run `aptos move publish`, Im getting `Simulation failed with status: Move abort in 0x1::resource_account: 0x60001"` error.

From that, I understand that the resource account was not found or that I used the wrong address in this line of code `let signer_cap = resource_account::retrieve_resource_account_cap(resource_signer, @source_addr);`.

Therefore I tried to [follow this](https://aptos.dev/en/build/smart-contracts/resource-accounts) in order to generate the address first using `create-resource-account-and-publish-package`like this `aptos move create-resource-account-and-publish-package --address-name aptos_institute --seed test` (aptos institute here is an address I got when I ran the aptos move compile). But I meet another error which is saying `"Error": "Unexpected error: Unable to resolve packages for package 'aptos_institute': Unable to resolve named address 'aptos_institute' in package 'aptos_institute' when resolving dependencies: Attempted to assign a different value '0x50a1c95f48059ee74cdd2dcc0b0eb60c0547ae37fef5b7231feb014a0c36baa1' to an a already-assigned named address '0x357017d43fab9fb7f46f097c69dc27cb1472ecddb28090f1652dbb1b3f604527'"`

Okay, so I try to use that new address I just got as the the address when I run `create-resource-account-and-publish-package`. That asks me if I really want to publish this package the resource account's address 0x357017d43fab9fb7f46f097c69dc27cb1472ecddb28090f1652dbb1b3f604527. I say yes. And get another error - `"Simulation failed with status: MODULE_ADDRESS_DOES_NOT_MATCH_SENDER"`.

I don't understand how am I suppose to publish with resource accounts at all. Could someone help me with what address to use? Or if I am doing something else wrong, point me to the mistake?

Thank you

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

Aptos CLI

**Comment by kshitijc1506** [comment]
I suspect the issue is coming from your `Move.toml`, you can use this e2e [example](https://github.com/aptos-labs/aptos-core/tree/main/aptos-move/move-examples/mint_nft/2-Using-Resource-Account) for reference.

**Reply by davd-ops** [reply]
Could you please point me a little closer? My Move.toml looks very similar to theirs in the example

**Reply by kshitijc1506** [reply]
There are two named addresses with the same address (`admin_addr` and `aptos_institute`), can you check this?

**Reply by davd-ops** [reply]
Correct that is because from what I understand the module needs to be deployed on the deployer address, so admin (for controlling some administration settings) is the same as deployer address (module deployment address).

**Reply by kshitijc1506** [reply]
I think it's just misconfiguration of your local environment. You can check your config.yaml to see your configured profile (sender) for this error:

"Simulation failed with status: `MODULE_ADDRESS_DOES_NOT_MATCH_SENDER`

**Reply by kshitijc1506** [reply]
1. Resource acc address should match Move.toml named address
2. If the address matches from the create RA and publish command, check if you have not already published a module with the same name to the same address

Lmk if both these things don't work out. Worst case, you can try re-configuring everything by running aptos init from scratch (delete existing config).

**Reply by davd-ops** [reply]
1. When I run the create RA command, it always tries to create the resource account under (`0x357017d43fab9fb7f46f097c69dc27cb1472ecddb28090f1652dbb1b3f604527`) address, when I did not specify that address in any way. I don't know why is it choosing this address. Do I need to fund this address with gas money?
2. I am a newbie on Aptos but based on what I checked, I don't see it published anywhere on the testnet explorer, neither on my address (`0x50a1c95f48059ee74cdd2dcc0b0eb60c0547ae37fef5b7231feb014a0c36baa1`), nor the assigned resource one. I do see [some resources](https://explorer.aptoslabs.com/account/0x50a1c95f48059ee74cdd2dcc0b0eb60c0547ae37fef5b7231feb014a0c36baa1/resources?network=testnet) under my account tho

Still didn't manage to fix it. I tried building another repo to test if this is maybe only a problem with this repo but I have the same issue there. Could you check if you clone my repo, if you get the same issue?

I appreciate it, thank you

**Reply by davd-ops** [reply]
I am thinking that when I run the create RA and publish command, it tries to do both things at the same time, and since the RA is used within my module, it fails? Is there a way to only deploy the resource account and publish after? Not sure if that makes sense but it's the reason I can think of

**Reply by davd-ops** [reply]
This si what im getting when I changed all addresses in Move.toml to be `0x357017d43fab9fb7f46f097c69dc27cb1472ecddb28090f1652dbb1b3f604527`
<img width="1326" alt="image" src="https://github.com/user-attachments/assets/b04eca6a-db9e-45c6-a929-a62c618251ad">

**Reply by davd-ops** [reply]
Okay I got some progress, I changed the init_module to not run on start but rather be a function that is called after like this
`public entry fun public_init_module(resource_signer: &signer, parent_address: address) {
    let signer_cap = resource_account::retrieve_resource_account_cap(resource_signer, parent_address);`
    
And I was able to execute the publish [in this TX](https://explorer.aptoslabs.com/txn/0x319d5b64c6a2f2ef3a35367350937d44c58787b08588bc97e30970faacdc3421?network=testnet) 
<img width="1330" alt="image" src="https://github.com/user-attachments/assets/e435e790-750c-42b1-9f6e-c40cd12dcde1">

So I think resource account should now be available? I am not sure if the module was also deployed or not I don't see it anywhere

If I now revert the changes to the `init_module` function and try to publish it using normal publish command, I am now getting a new error `"Error": "Simulation failed with status: Move abort in 0x1::resource_account: 0x10002"`

**Reply by kshitijc1506** [reply]
@davd-ops I cloned, so basically:

1. Remove `source_addr` from your Move.toml entirely (this is your `admin_addr`, AKA the account that creates the resource account) -- unless you want them to be different but I don't think that's the case. 
2. Make the `admin_addr` the same as the account address in your `config.yaml` (I think this is already the case)
3. Change the named address for `aptos_institute` to a placeholder - the CLI needs this since `@aptos_institute` represents the address of the resource account after publishing. 
```
    [addresses]
    admin_addr = "0x88b2d8344be13ca52da70764f7cb86c14bbae26e63834a65c407c182fa96d4f4"
    aptos_institute="_"
```

In your `init_module`, you need to change your `retrieve_resource_account_cap` to this:

`let signer_cap = resource_account::retrieve_resource_account_cap(resource_signer, @admin_addr);
`

**Reply by kshitijc1506** [reply]
This solves the resource account issue, but, there are other runtime errors popping up in `init_module`. I suggest writing a few tests to utilize the stack trace and find out exactly where the error is coming from.

**Comment by kshitijc1506** [comment]
Also, I recommend you use objects for deployment and contract state storage, it's a lot easier to deal with :)
